slo_generator/backends/cloud_monitoring_mql.py

```diff
@@ -0,0 +1,230 @@
+        threshold_bucket: int = int(measurement['threshold_bucket'])
+        good_below_threshold: bool = measurement.get('good_below_threshold',
+                                                     True)
+
+        # Query 'valid' events
+        series = self.query(query=filter_valid, window=window)
+
+        if not series:
+            return NO_DATA, NO_DATA  # no timeseries
+
+        # FIXME Update path to distribution value
+        distribution_value = series[0].point_data[0].values[
+            0].distribution_value
+        # bucket_options = distribution_value.bucket_options
+        bucket_counts = distribution_value.bucket_counts
+        valid_events_count = distribution_value.count
+        # growth_factor = bucket_options.exponential_buckets.growth_factor
+        # scale = bucket_options.exponential_buckets.scale
+
+        # Explicit the exponential distribution result
+        count_sum = 0
+        distribution = OrderedDict()
+        for i, bucket_count in enumerate(bucket_counts):
+            count_sum += bucket_count
+            # upper_bound = scale * math.pow(growth_factor, i)
+            distribution[i] = {
+                # 'upper_bound': upper_bound,
+                # 'bucket_count': bucket_count,
+                'count_sum': count_sum
+            }
+        LOGGER.debug(pprint.pformat(distribution))
+
+        if len(distribution) - 1 < threshold_bucket:
+            # maximum measured metric is below the cut after bucket number
+            lower_events_count: int = valid_events_count
+            upper_events_count: int = 0
+        else:
+            lower_events_count: int = distribution[threshold_bucket][
+                'count_sum']
+            upper_events_count: int = valid_events_count - lower_events_count
+
+        if good_below_threshold:
+            good_event_count: int = lower_events_count
+            bad_event_count: int = upper_events_count
+        else:
+            good_event_count: int = upper_events_count
+            bad_event_count: int = lower_events_count
+
+        return good_event_count, bad_event_count
+
+    def exponential_distribution_cut(self, *args, **kwargs) -> tuple[int, int]:
+        """Alias for `distribution_cut` method to allow for backwards
+        compatibility.
+        """
+        warnings.warn(
+            'exponential_distribution_cut will be deprecated in version 2.0, '
+            'please use distribution_cut instead', FutureWarning)
+        return self.distribution_cut(*args, **kwargs)
+
+    def query_sli(self, timestamp: int, window: int, slo_config: dict) -> float:
+        """Query SLI value from a given MQL query.
+
+        Args:
+            timestamp (int): UNIX timestamp.
+            window (int): Window (in seconds).
+            slo_config (dict): SLO configuration.
+
+        Returns:
+            float: SLI value.
+        """
+        # FIXME This method returns 2.0 for the 12-hour, 7-day and 28-day burn
```

Comment:
Not sure what is going wrong here. I will investigate further over the next few days. This is still a WIP.

Reply:
I found the root cause. I need to aggregate (sum) the time series **before** computing the `ratio`.